File: pos-frontend/electron/services/tenantSeed.ts
import bcrypt from "bcryptjs";
import { existsSync, mkdirSync, writeFileSync } from "fs";
import { dirname, join } from "path";

/**
 * First-run tenant seeding. The launcher collects the business name + admin
 * credentials, bcrypt-hashes the password HERE (so plaintext never lands on
 * disk), and writes tenant-seed.json for the backend's DesktopBootstrapRunner
 * to consume once on first boot.
 */

export interface FirstRunInput {
  tenantName: string;
  adminEmail: string;
  adminPassword: string;
}

/** Must match the path main.ts passes to the backend as APP_TENANT_SEED_FILE. */
export function tenantSeedPath(userDataDir: string): string {
  return join(userDataDir, "config", "tenant-seed.json");
}

/**
 * True when no seed has been written yet. Note: if the DB is later wiped but the
 * seed file remains, the backend re-seeds from it (its tenant-count guard) using
 * the same credentials — so we deliberately do NOT re-prompt in that case.
 */
export function needsFirstRun(userDataDir: string): boolean {
  return !existsSync(tenantSeedPath(userDataDir));
}

export function writeTenantSeed(userDataDir: string, input: FirstRunInput): void {
  const path = tenantSeedPath(userDataDir);
  mkdirSync(dirname(path), { recursive: true });
  const seed = {
    tenantName: input.tenantName.trim(),
    adminEmail: input.adminEmail.trim().toLowerCase(),
    // cost 10 — Spring's BCryptPasswordEncoder verifies the $2a$/$2b$ output.
    adminPasswordBcrypt: bcrypt.hashSync(input.adminPassword, 10),
  };
  writeFileSync(path, JSON.stringify(seed, null, 2), { encoding: "utf8" });
}

File: pos-backend/src/main/java/com/lumora/pos/superadmin/dto/TenantSeed.java
package com.lumora.pos.superadmin.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;

/**
 * First-run tenant seed written by the Electron desktop launcher to
 * {@code %LOCALAPPDATA%\LumoraPOS\config\tenant-seed.json}. The admin password
 * arrives ALREADY bcrypt-hashed — the launcher hashes it so plaintext never
 * touches disk. Consumed once by {@code DesktopBootstrapRunner} on first boot.
 */
@JsonIgnoreProperties(ignoreUnknown = true)
public record TenantSeed(
        String tenantName,
        String adminEmail,
        String adminPasswordBcrypt
) {}
